File: part1/node/MyBot.js
'use strict';

const ConversationV1 = require('watson-developer-cloud/conversation/v1');

class MyBot {

    /**
     * Creates a new instance of MyBot.
     * @param {string} conversationUsername - The Watson Conversation username
     * @param {string} conversationPassword - The Watson Converation password
     * @param {string} conversationWorkspaceId - The Watson Conversation workspace ID
     */
    constructor(conversationUsername, conversationPassword, conversationWorkspaceId) {
        this.conversationService = new ConversationV1({
            username: conversationUsername,
            password: conversationPassword,
            version_date: '2017-04-21'
        });
        this.conversationWorkspaceId = conversationWorkspaceId;
        this.conversationContext = null;
    }

    /**
     * Process the message entered by the user.
     * @param {string} message - The message entered by the user
     * @returns {Promise.<string|Error>} - The reply to be sent to the user if fulfilled, or an error if rejected
     */
    processMessage(message) {
        // The first step is to send the message entered by the user to Watson Conversation.
        // We send the conversationContext associated with the current user.
        // In this application there is only a single user, so we use the global conversationContext variable.
        // In a typical application you would associate the context with a user, and whenever
        // a new message is received you would look up that user and their context based on the User ID
        // from the messaging platform (for example, the Slack ID).
        let conversationResponse = null;
        return this.sendRequestToWatsonConversation(message, this.conversationContext)
            .then((response) => {
                conversationResponse = response;
                return this.handleResponseFromWatsonConversation(conversationResponse);
            })
            .then((reply) => {
                // Update our local conversationContext every time we receive a response from Watson Conversation.
                // This keeps track of the active dialog in the conversation.
                this.conversationContext = conversationResponse.context;
                // Return the reply to be sent to the user.
                return Promise.resolve(reply);
            })
            .catch((error) => {
                console.log(`Error: ${JSON.stringify(error,null,2)}`);
                let reply = 'Sorry, something went wrong!\n'
                return Promise.resolve(reply);
            });
    }

    /**
     * Sends the message entered by the user to Watson Conversation
     * along with the active Watson Conversation context that is used to keep track of the conversation.
     * @param {string} message - The message entered by the user
     * @param {object} conversationContext - The active Watson Conversation context
     * @returns {Promise.<object|error>} - The response from Watson Conversation if fulfilled, or an error if rejected
     */
    sendRequestToWatsonConversation(message, conversationContext) {
        return new Promise((resolve, reject) => {
            var conversationRequest = {
                input: {text: message},
                context: conversationContext,
                workspace_id: this.conversationWorkspaceId,
            };
            this.conversationService.message(conversationRequest, (error, response) => {
                if (error) {
                    reject(error);
                }
                else {
                    resolve(response);
                }
            });
        });
    }
    
    /**
     * Takes the response from Watson Conversation, performs any additional steps
     * that may be required, and returns the reply that should be sent to the user.
     * @param {object} conversationResponse - The response from Watson Conversation
     * @returns {Promise.<string|error>} - The reply to send to the user if fulfilled, or an error if rejected
     */
    handleResponseFromWatsonConversation(conversationResponse) {
        // In some cases we just return the response defined in Watson Conversation (handled by handleDefaultMessage).
        // In others we need to take special steps to return a customized response.
        // For example, we may need to return the results of a databse lookup or 3rd party API call.
        // Here we look to see if a custom "action" has been configured in Watson Conversation and if we
        // need to return a custom response based on the action. 
        const action = conversationResponse.context.action;
        if (action == "xxx") {
             return this.handleXXXMessage(conversationResponse);
        }
        else {
            return this.handleDefaultMessage(conversationResponse);
        }
    }

    /**
     * The default handler for any message from Watson Conversation that requires no additional steps.
     * Returns the reply that was configured in the Watson Conversation dialog.
     * @param {object} conversationResponse - The response from Watson Conversation
     * @returns {Promise.<string|error>} - The reply to send to the user if fulfilled, or an error if rejected
     */
    handleDefaultMessage(conversationResponse) {
        let reply = '';
        for (let i = 0; i < conversationResponse.output.text.length; i++) {
            reply += conversationResponse.output.text[i] + '\n';
        }
        return Promise.resolve(reply);
    }

    /**
     * Returns a custom response to the user.
     * @param {object} conversationResponse - The response from Watson Conversation
     * @returns {Promise.<string|error>} - The reply to send to the user if fulfilled, or an error if rejected
     */
    handleXXXMessage(conversationResponse) {
        // let entity = conversationResponse.entities[0].value
        // let contextVar = conversationResponse.context.var
        let reply = 'TBD';
        return Promise.resolve(reply);
    }
}

module.exports = MyBot;

File: part1/python/my_bot.py
import sys

from watson_developer_cloud import ConversationV1

class MyBot():

    def __init__(self, conversation_username, conversation_password, conversation_workspace_id):
        """
        Creates a new instance of MyBot.
        Parameters
        ----------
        conversation_username - The Watson Conversation username
        conversation_password - The Watson Converation password
        conversation_workspace_id - The Watson Conversation workspace ID
        """
        self.conversation_client = ConversationV1(
            username=conversation_username,
            password=conversation_password,
            version='2016-07-11'
        )
        self.conversation_workspace_id = conversation_workspace_id
        self.conversation_context = None

    def process_message(self, message):
        """
        Process the message entered by the user.
        Parameters
        ----------
        message - The message entered by the user
        """
        conversation_response = None
        try:
            conversation_response = self.send_request_to_watson_conversation(message, self.conversation_context)
            reply = self.handle_response_from_watson_conversation(message, conversation_response)
            self.conversation_context = conversation_response['context']
            return reply
        except Exception:
            print(sys.exc_info())
            # clear state and set response
            reply = "Sorry, something went wrong!"
            return reply

    def send_request_to_watson_conversation(self, message, conversation_context):
        """
        Sends the message entered by the user to Watson Conversation
        along with the active Watson Conversation context that is used to keep track of the conversation.
        Parameters
        ----------
        message - The message entered by the user
        conversation_context - The active Watson Conversation context
        """
        return self.conversation_client.message(
            workspace_id=self.conversation_workspace_id,
            message_input={'text': message},
            context=conversation_context
        )

    def handle_response_from_watson_conversation(self, message, conversation_response):
        """ 
        Takes the response from Watson Conversation, performs any additional steps
        that may be required, and returns the reply that should be sent to the user.
        Parameters
        ----------
        message - The message sent by the user
        conversation_response - The response from Watson Conversation
        """
            
        # In some cases we just return the response defined in Watson Conversation (handled by handleDefaultMessage).
        # In others we need to take special steps to return a customized response.
        # For example, we may need to return the results of a databse lookup or 3rd party API call.
        # Here we look to see if a custom "action" has been configured in Watson Conversation and if we
        # need to return a custom response based on the action. 
        if 'context' in conversation_response.keys() and 'action' in conversation_response['context'].keys():
            action = conversation_response['context']['action']
        else:
            action = None
        if action == "xxx":
            return self.handle_xxx_message(conversation_response)
        else:
            return self.handle_default_message(conversation_response)
        
    def handle_default_message(self, conversation_response):
        """
        The default handler for any message from Watson Conversation that requires no additional steps.
        Returns the reply that was configured in the Watson Conversation dialog.
        Parameters
        ----------
        conversation_response - The response from Watson Conversation
        """
        reply = ''
        for text in conversation_response['output']['text']:
            reply += text + "\n"
        return reply

    def handle_xxx_message(self, conversation_response):
        """
        Returns a custom response to the user.
        Parameters
        ----------
        conversation_response - The response from Watson Conversation
        """
        # entity = conversation_response['entities'][0]['value']
        # contextVar = conversation_response['context']['var']
        reply = 'TBD'
        return reply